places/src/db/db.rs

```diff
@@ -0,0 +1,274 @@
+        } else {
+            self.db.execute_named(sql, params)
+        };
+        if let Err(e) = &res {
+            warn!("Error running SQL {}. Statement: {:?}", e, sql);
+        }
+        Ok(res?)
+    }
+
+    pub fn query_one<T: FromSql>(&self, sql: &str) -> Result<T> {
+        let res: T = self.db.query_row(sql, &[], |row| row.get(0))?;
+        Ok(res)
+    }
+
+    // Note that there are several differences between these and `self.db.query_row`: it returns
+    // None and not an error if no rows are returned, it allows the function to return a result, etc
+    pub fn query_row_cached<T>(&self, sql: &str, args: &[&ToSql], f: impl FnOnce(&Row) -> Result<T>) -> Result<Option<T>> {
+        let mut stmt = self.db.prepare_cached(sql)?;
+        let res = stmt.query(args);
+        if let Err(e) = &res {
+            warn!("Error executing query: {}. Query: {}", e, sql);
+        }
+        let mut rows = res?;
+        match rows.next() {
+            Some(result) => Ok(Some(f(&result?)?)),
+            None => Ok(None),
+        }
+    }
+
+    // cached and uncached stmt types are completely different so we can't remove the duplication
+    // between query_row_cached and query_row... :/
+    pub fn query_row<T>(&self, sql: &str, args: &[&ToSql], f: impl FnOnce(&Row) -> Result<T>) -> Result<Option<T>> {
+        let mut stmt = self.db.prepare(sql)?;
+        let res = stmt.query(args);
+        if let Err(e) = &res {
+            warn!("Error executing query: {}. Query: {}", e, sql);
+        }
+        let mut rows = res?;
+        match rows.next() {
+            Some(result) => Ok(Some(f(&result?)?)),
+            None => Ok(None),
+        }
+    }
+
+    pub fn query_row_named<T>(&self, sql: &str, args: &[(&str, &ToSql)], f: impl FnOnce(&Row) -> Result<T>) -> Result<Option<T>> {
+        let mut stmt = self.db.prepare(sql)?;
+        let res = stmt.query_named(args);
+        if let Err(e) = &res {
+            warn!("Error executing query: {}. Query: {}", e, sql);
+        }
+        let mut rows = res?;
+        match rows.next() {
+            Some(result) => Ok(Some(f(&result?)?)),
+            None => Ok(None),
+        }
+    }
+}
+
+// ----------------------------- end of stuff that should be common --------------------
+
+fn define_functions(c: &Connection) -> Result<()> {
+    c.create_scalar_function("autocomplete_match", -1, true, move |ctx| {
+        let search_string = ctx.get::<Option<String>>(0)?.unwrap_or_default();
+        let url = ctx.get::<Option<String>>(1)?.unwrap_or_default();
+        let title = ctx.get::<Option<String>>(2)?.unwrap_or_default();
+        let tags = ctx.get::<Option<String>>(3)?;
+        let visit_count = ctx.get::<i64>(4)?;
+        let typed = ctx.get::<bool>(5)?;
+        let bookmarked = ctx.get::<bool>(6)?;
+        let open_page_count = ctx.get::<Option<i64>>(7)?;
+        let match_behavior = ctx.get::<MatchBehavior>(8);
+
+        let trimmed_url = &url[..255];
```

Comment:
This will panic for short urls. You need to do `&url[..url.len().min(255)]`. I agree this is annoying.